ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -1278,13 +1282,100 @@ private class FirstGetChunk extends GetChunk {
+      this.blobMetadataCache = blobMetadataCache;
+      if (shouldLookupMetadataCache()) {
+        BlobMetadata blobMetadata = (BlobMetadata) blobMetadataCache.getObject(blobId.toString());
+        if (blobMetadata != null) {
+          blobMetadataCacheEntryStatus = AmbryCacheEntryStatus.cacheHit;
+          blobType = BlobType.MetadataBlob;
+          compositeBlobInfo = blobMetadata.getCompositeBlobInfo();
+          blobInfo = blobMetadata.getBlobInfo();
+          lifeVersion = blobInfo.getLifeVersion();
+          serverBlobProperties = blobInfo.getBlobProperties();
+          chunkIndexToBuf = new ConcurrentHashMap<>();
+          chunkIndexToBufWaitingForRelease = new ConcurrentHashMap<>();
+          successfullyDeserialized = true;
+        } else {
+          blobMetadataCacheEntryStatus = AmbryCacheEntryStatus.cacheMiss;
+        }
+      }
+    }
+
+    /**
+     * Decides if we should look up the metadata cache during a GET request
+     * @return True if and only if
+     *  - cache exists
+     *  - it is a RANGE request
+     *  - it is not a raw mode request
+     *  - the blobID is for a metadata blob
+     *  - the cache not yet been consulted
+     *  False, otherwise.
+     */
+    boolean shouldLookupMetadataCache() {
+      boolean result = (blobMetadataCache != null &&
+          options.getBlobOptions.getRange() != null &&
+          options.getBlobOptions.isRawMode() == false &&
+          blobId.getBlobDataType() == BlobId.BlobDataType.METADATA &&
+          blobMetadataCacheEntryStatus == AmbryCacheEntryStatus.cacheNotLookedUp);
+      return result;
+    }
+
+    /**
+     * Decides if we should cache metadata during a GET request
+     * This method should be called after decryption is complete.
+     * @return True if and only if
+     *  - cache exists
+     *  - it is RANGE request
+     *  - it is not a raw mode request
+     *  - the blobID is for a metadata blob
+     *  - the cache was looked up and we got a cache miss
+     *  False, otherwise.
+     */
+    boolean shouldSaveMetadata() {
+      boolean result = (blobMetadataCache != null &&
+          options.getBlobOptions.getRange() != null &&
+          options.getBlobOptions.isRawMode() == false &&
+          blobId.getBlobDataType() == BlobId.BlobDataType.METADATA &&
+          blobMetadataCacheEntryStatus == AmbryCacheEntryStatus.cacheMiss &&
+          totalSize >= routerConfig.routerSmallestBlobForMetadataCache);
+      return result;
+    }
+
+    /**
+     * Conditionally saves blob metadata for composite blobs
+     * @return True if metadata was saved successfully, else False.
+     */
+    boolean saveMetadata() {
+      boolean result = false;
+      if (shouldSaveMetadata()) {
+        BlobMetadata blobMetadata = new BlobMetadata(blobId.toString(), blobInfo, compositeBlobInfo);
+        result = blobMetadataCache.putObject(blobMetadata.getBlobId(), blobMetadata);
+        if (result == true) {
```

Comment:
if (result) {
}

Reply:
actually, no. The result var is the result of inserting an object in the cache and not the result of shouldSaveMetadata. So this code is correct.